staging/src/k8s.io/client-go/applyconfigurations/meta/v1/unstructured.go

```diff
@@ -0,0 +1,130 @@
+package v1
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/managedfields"
+	"k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/util/proto"
+	"sigs.k8s.io/structured-merge-diff/v4/typed"
+)
+
+// UnstructuredExtractor enables extracting the applied configuration state from object for fieldManager into an
+// unstructured object type.
+type UnstructuredExtractor interface {
+	ExtractUnstructured(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error)
+	ExtractUnstructuredStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error)
+}
+
+// gvkParserCache caches the GVKParser in order to prevent from having to repeatedly
+// parse the models from the open API schema when the schema itself changes infrequently.
+type gvkParserCache struct {
+	// discoveryClient is the client for retrieving the openAPI document and checking
+	// whether the document has changed recently
+	discoveryClient discovery.DiscoveryInterface
+	// ttl is how long the openAPI schema should be considered valid
+	ttl time.Duration
+	// mu protects the gvkParser
+	mu sync.Mutex
+	// gvkParser retrieves the objectType for a given gvk
+	gvkParser *fieldmanager.GvkParser
+	// lastChecked is the last time we checked if the openAPI doc has changed.
+	lastChecked time.Time
+}
+
+// regenerateGVKParser builds the parser from the raw OpenAPI schema.
+func (c *gvkParserCache) regenerateGVKParser() error {
+	doc, err := c.discoveryClient.OpenAPISchema()
```

Comment:
I would consider modifying this method to also return the etag (or make a parallel method that does so), so that you have a key you can use in any caches.

Perhaps also modify it to take in a prior etag, so that it can avoid any internal parsing involved in producing `doc`.

Reply:
My opinion is that the cache mechanism is mostly hidden from you (or should), within gregjones. You don't really *know* what etag you have available, the httpcache layer knows much better.